File: python/graph.py
from abc import ABC, abstractmethod
from typing import TypeVar, Generic, Callable, Dict, Set, List, Optional, Union, Any, Iterator

NodeType = TypeVar('NodeType')
WeightType = TypeVar('WeightType')

class GraphException(Exception):
    """Exception class for Graph-related errors."""
    pass


class Node(Generic[NodeType]):
    """
    Represents a Node (Vertex) in the graph.

    Attributes:
        value (NodeType): The internal data stored inside the node.
    """

    def __init__(self, value: NodeType):
        """
        Initializes a new node with the given value.

        Args:
            value (NodeType): The data to be stored in the node.
        """
        self._value = value

    @property
    def value(self) -> NodeType:
        """NodeType: Gets the internal value of the node."""
        return self._value

    def __str__(self) -> str:
        """Returns the string representation of the node's value."""
        return str(self._value)


class Arc(Generic[NodeType, WeightType]):
    """
    Represents a directed or undirected Arc (Edge) between two nodes.

    Attributes:
        source (Node[NodeType]): The starting node of the arc.
        target (Node[NodeType]): The ending node of the arc.
    """

    def __init__(self,
                 source: Node[NodeType],
                 target: Node[NodeType],
                 weight: Union[WeightType, Callable[[], WeightType], None] = None):
        """
        Constructs an arc. If weight is a callable (lambda), it evaluates dynamically.
        If it's a fixed value, it wraps it in a lambda automatically to maintain closures logic.

        Args:
            source (Node[NodeType]): The origin node.
            target (Node[NodeType]): The destination node.
            weight (Union[WeightType, Callable, None]): The weight of the arc, a dynamic function, or None.
        """
        self._source = source
        self._target = target

        if callable(weight):
            self._weight_function = weight
        else:
            self._weight_function = lambda: weight

    @property
    def source(self) -> Node[NodeType]:
        """Node[NodeType]: Retrieves the source node."""
        return self._source

    @property
    def target(self) -> Node[NodeType]:
        """Node[NodeType]: Retrieves the target node."""
        return self._target

    def get_weight(self) -> Optional[WeightType]:
        """
        Evaluates and returns the current weight of the arc.

        Returns:
            Optional[WeightType]: The evaluated weight, or None if the graph is unweighted.
        """
        return self._weight_function()

    def set_weight_function(self, new_weight_func: Callable[[], WeightType]) -> None:
        """
        Updates the dynamic weight evaluation function.

        Args:
            new_weight_func (Callable[[], WeightType]): The new lambda function returning a weight.
        """
        self._weight_function = new_weight_func

    def print_arc(self, is_digraph: bool) -> None:
        """
        Prints the arc information to standard output.

        Args:
            is_digraph (bool): True if directed formatting should be used, False for undirected.
        """
        connection = "From: " if is_digraph else "Union With Nodes: "
        direction = " To: " if is_digraph else " And: "

        base_str = f"{connection}{self.source.value}{direction}{self.target.value}"

        weight = self.get_weight()
        if weight is not None:
            print(f"{base_str} Weight: {weight}")
        else:
            print(base_str)


class Graph(ABC, Generic[NodeType, WeightType]):
    """
    Abstract Base Class for Graphs.
    Contains the shared state and common persistence logic for all graph types.
    """

    def __init__(self):
        """Initializes an empty graph with internal data structures."""
        self._nodes: Set[Node[NodeType]] = set()
        self._arcs: Set[Arc[NodeType, WeightType]] = set()

        # Adjacency lists are critical for the O((V+E) log V) time complexity in algorithms.
        self._out_adj_list: Dict[Node[NodeType], List[Arc[NodeType, WeightType]]] = {}
        self._in_adj_list: Dict[Node[NodeType], List[Arc[NodeType, WeightType]]] = {}

    # --- Python Data Model Magic Methods ---

    def __len__(self) -> int:
        """Returns the total number of nodes in the graph in O(1) time."""
        return len(self._nodes)

    def __contains__(self, node: Node[NodeType]) -> bool:
        """Checks if a specific node exists within the graph in O(1) time."""
        return node in self._nodes

    def __iter__(self) -> Iterator[Node[NodeType]]:
        """Allows pythonic iteration over all nodes in the graph (e.g., 'for node in graph:')."""
        return iter(self._nodes)

    # ---------------------------------------

    def add_node_by_value(self, value: NodeType) -> Node[NodeType]:
        """
        Creates a new node by value and inserts it into the graph.

        Args:
            value (NodeType): The data for the new node.

        Returns:
            Node[NodeType]: The newly created and inserted node.
        """
        node = Node(value)
        self._nodes.add(node)
        self._out_adj_list[node] = []
        self._in_adj_list[node] = []
        return node

    def add_node(self, node: Node[NodeType]) -> None:
        """
        Inserts an already existing node instance into the graph.

        Args:
            node (Node[NodeType]): The node to insert.
        """
        self._nodes.add(node)
        if node not in self._out_adj_list:
            self._out_adj_list[node] = []
        if node not in self._in_adj_list:
            self._in_adj_list[node] = []

    @abstractmethod
    def add_arc(self, source: Node[NodeType], target: Node[NodeType], weight: Union[WeightType, Callable[[], WeightType], None] = None) -> Arc[NodeType, WeightType]:
        """Pure virtual method to enforce implementation in derived classes."""
        pass

    @abstractmethod
    def print_graph(self) -> None:
        """Pure virtual method to print the graph structure."""
        pass

    @abstractmethod
    def to_dot_file(self, filename: str) -> None:
        """Pure virtual method to export the graph to a Graphviz DOT file."""
        pass

    @property
    def nodes(self) -> Set[Node[NodeType]]:
        """Set[Node]: Retrieves the set of all nodes in the graph."""
        return self._nodes

    @property
    def arcs(self) -> Set[Arc[NodeType, WeightType]]:
        """Set[Arc]: Retrieves the set of all arcs in the graph."""
        return self._arcs

    def get_outgoing_arcs(self, node: Node[NodeType]) -> List[Arc[NodeType, WeightType]]:
        return self._out_adj_list.get(node, [])

    def get_incoming_arcs(self, node: Node[NodeType]) -> List[Arc[NodeType, WeightType]]:
        return self._in_adj_list.get(node, [])

    def get_out_degree(self, node: Node[NodeType]) -> int:
        return len(self.get_outgoing_arcs(node))

    def get_in_degree(self, node: Node[NodeType]) -> int:
        return len(self.get_incoming_arcs(node))

    def get_degree(self, node: Node[NodeType]) -> int:
        return self.get_out_degree(node) + self.get_in_degree(node)

    def save_to_file(self, filename: str) -> None:
        """
        Persists the graph's current state into a text file.
        Follows a sequential serialization format compatible with the C++ implementation.

        Args:
            filename (str): The destination file path.

        Raises:
            GraphException: If the file cannot be opened for writing.
        """
        try:
            with open(filename, 'w') as file:
                # 1. Write total number of nodes
                file.write(f"{len(self._nodes)}\n")

                # 2. Map nodes to integer indices for O(1) serialization
                node_to_index: Dict[Node[NodeType], int] = {}
                ordered_nodes = list(self._nodes)

                for index, node in enumerate(ordered_nodes):
                    node_to_index[node] = index
                    file.write(f"{node.value}\n")

                # 3. Write total number of arcs
                file.write(f"{len(self._arcs)}\n")

                # 4. Write arc connections
                for arc in self._arcs:
                    src_idx = node_to_index[arc.source]
                    tgt_idx = node_to_index[arc.target]
                    weight = arc.get_weight()

                    # Compile-time evaluation equivalent: Only write weight if it exists
                    if weight is not None:
                        file.write(f"{src_idx} {tgt_idx} {weight}\n")
                    else:
                        file.write(f"{src_idx} {tgt_idx}\n")
        except IOError:
            raise GraphException("Could not open file for writing.")

    @classmethod
    def load_from_file(cls, filename: str, value_cast: Callable[[str], Any] = str, weight_cast: Callable[[str], Any] = float) -> 'Graph':
        """
        Instantiates a new graph by deserializing a text file.
        Fully compatible with files generated by the C++ Graph library.

        Args:
            filename (str): The source file path.
            value_cast (Callable): Function to cast node values from string (e.g., int, str).
            weight_cast (Callable): Function to cast arc weights from string (e.g., float, int).

        Returns:
            Graph: The reconstructed DirectedGraph or UndirectedGraph.

        Raises:
            GraphException: If the file cannot be opened for reading.
        """
        g = cls() # Factory pattern: Instantiates DirectedGraph or UndirectedGraph automatically
        try:
            with open(filename, 'r') as file:
                tokens = file.read().split()
                if not tokens:
                    return g

                token_idx = 0
                num_nodes = int(tokens[token_idx])
                token_idx += 1

                index_to_node: List[Node[NodeType]] = []
                for _ in range(num_nodes):
                    val = value_cast(tokens[token_idx])
                    token_idx += 1
                    index_to_node.append(g.add_node_by_value(val))

                if token_idx >= len(tokens):
                    return g

                num_arcs = int(tokens[token_idx])
                token_idx += 1

                # Token counting to infer if the graph is weighted
                # Total remaining tokens / number of arcs. 3 means weighted (src, tgt, weight), 2 means unweighted.
                tokens_per_arc = (len(tokens) - token_idx) // num_arcs if num_arcs > 0 else 0

                for _ in range(num_arcs):
                    src_idx = int(tokens[token_idx])
                    tgt_idx = int(tokens[token_idx+1])

                    if tokens_per_arc >= 3:
                        weight_val = weight_cast(tokens[token_idx+2])
                        g.add_arc(index_to_node[src_idx], index_to_node[tgt_idx], weight_val)
                    else:
                        g.add_arc(index_to_node[src_idx], index_to_node[tgt_idx], None)

                    token_idx += tokens_per_arc

        except IOError:
            raise GraphException("Could not open file for reading.")

        return g


class DirectedGraph(Graph[NodeType, WeightType]):
    """Implementation of a Directed Graph structure."""

    def add_arc(self, source: Node[NodeType], target: Node[NodeType], weight: Union[WeightType, Callable[[], WeightType], None] = None) -> Arc[NodeType, WeightType]:
        """
        Adds a directed arc from the source node to the target node.

        Raises:
            GraphException: If the source or target node is not present in the graph.
        """
        if source not in self._nodes or target not in self._nodes:
            raise GraphException("Nodes Not Found!")

        arc = Arc(source, target, weight)
        self._arcs.add(arc)

        # Directed: The edge is outgoing from source and incoming to target.
        self._out_adj_list[source].append(arc)
        self._in_adj_list[target].append(arc)

        return arc

    def print_graph(self) -> None:
        """Prints all directed arcs in the graph."""
        for arc in self._arcs:
            arc.print_arc(is_digraph=True)

    def to_dot_file(self, filename: str) -> None:
        """
        Generates a DOT file for Graphviz visualization of a directed graph.

        Args:
            filename (str): The output file path (e.g., 'graph.dot').

        Raises:
            GraphException: If the file cannot be written.
        """
        try:
            with open(filename, 'w') as file:
                file.write("digraph G {\n")
                for node in self._nodes:
                    file.write(f'  "{node.value}";\n')

                for arc in self._arcs:
                    weight = arc.get_weight()
                    if weight is not None:
                        file.write(f'  "{arc.source.value}" -> "{arc.target.value}" [label="{weight}"];\n')
                    else:
                        file.write(f'  "{arc.source.value}" -> "{arc.target.value}";\n')
                file.write("}\n")
        except IOError:
            raise GraphException("The file could not be opened for writing.")


class UndirectedGraph(Graph[NodeType, WeightType]):
    """Implementation of an Undirected Graph structure."""

    def add_arc(self, source: Node[NodeType], target: Node[NodeType], weight: Union[WeightType, Callable[[], WeightType], None] = None) -> Arc[NodeType, WeightType]:
        """
        Adds an undirected arc (bidirectional connection) between source and target nodes.

        Raises:
            GraphException: If the source or target node is not present in the graph.
        """
        if source not in self._nodes or target not in self._nodes:
            raise GraphException("Nodes Not Found!")

        arc = Arc(source, target, weight)
        self._arcs.add(arc)

        # Undirected: The edge is incident to both source and target.
        self._out_adj_list[source].append(arc)
        self._in_adj_list[source].append(arc)
        self._out_adj_list[target].append(arc)
        self._in_adj_list[target].append(arc)

        return arc

    def print_graph(self) -> None:
        """Prints all undirected connections in the graph."""
        for arc in self._arcs:
            arc.print_arc(is_digraph=False)

    def to_dot_file(self, filename: str) -> None:
        """
        Generates a DOT file for Graphviz visualization of an undirected graph.

        Args:
            filename (str): The output file path (e.g., 'graph.dot').

        Raises:
            GraphException: If the file cannot be written.
        """
        try:
            with open(filename, 'w') as file:
                file.write("graph G {\n")
                for node in self._nodes:
                    file.write(f'  "{node.value}";\n')

                for arc in self._arcs:
                    weight = arc.get_weight()
                    if weight is not None:
                        file.write(f'  "{arc.source.value}" -- "{arc.target.value}" [label="{weight}"];\n')
                    else:
                        file.write(f'  "{arc.source.value}" -- "{arc.target.value}";\n')
                file.write("}\n")
        except IOError:
            raise GraphException("The file could not be opened for writing.")

File: rust/graph_lib/src/graph.rs
/// @file graph.rs
/// @brief Core graph data structures and traits.

use std::cell::RefCell;
use std::collections::{HashMap, HashSet};
use std::fmt::Display;
use std::fs::File;
use std::hash::{Hash, Hasher};
use std::io::{Read, Write};
use std::rc::Rc;
use std::str::FromStr;
use std::sync::atomic::{AtomicUsize, Ordering};

static ID_COUNTER: AtomicUsize = AtomicUsize::new(0);

fn next_id() -> usize {
    ID_COUNTER.fetch_add(1, Ordering::SeqCst)
}

/// Represents a Node (Vertex) in the graph.
#[derive(Debug)]
pub struct Node<N> {
    pub id: usize,
    pub value: N,
}

impl<N> Node<N> {
    pub fn new(value: N) -> Self {
        Self { id: next_id(), value }
    }
}

impl<N> Hash for Node<N> {
    fn hash<H: Hasher>(&self, state: &mut H) { self.id.hash(state); }
}
impl<N> PartialEq for Node<N> {
    fn eq(&self, other: &Self) -> bool { self.id == other.id }
}
impl<N> Eq for Node<N> {}

/// Represents a directed or undirected Arc (Edge) between two nodes.
pub struct Arc<N, W> {
    pub id: usize,
    pub source: Rc<Node<N>>,
    pub target: Rc<Node<N>>,
    weight_function: RefCell<Option<Box<dyn Fn() -> W>>>,
}

impl<N, W> Arc<N, W> {
    pub fn new(source: Rc<Node<N>>, target: Rc<Node<N>>, weight_func: Option<Box<dyn Fn() -> W>>) -> Self {
        Self { id: next_id(), source, target, weight_function: RefCell::new(weight_func) }
    }

    pub fn new_fixed(source: Rc<Node<N>>, target: Rc<Node<N>>, weight: W) -> Self
    where W: Clone + 'static, {
        Self { id: next_id(), source, target, weight_function: RefCell::new(Some(Box::new(move || weight.clone()))) }
    }

    pub fn new_unweighted(source: Rc<Node<N>>, target: Rc<Node<N>>) -> Self {
        Self { id: next_id(), source, target, weight_function: RefCell::new(None) }
    }

    /// Evaluates and returns the current weight of the arc.
    pub fn get_weight(&self) -> Option<W> {
        let func = self.weight_function.borrow();
        func.as_ref().map(|f| f())
    }

    pub fn print(&self, is_digraph: bool) where N: Display, W: Display {
        let connection = if is_digraph { "From: " } else { "Union With Nodes: " };
        let direction = if is_digraph { " To: " } else { " And: " };

        print!("{}{}{}{}", connection, self.source.value, direction, self.target.value);
        if let Some(w) = self.get_weight() { println!(" Weight: {}", w); } else { println!(); }
    }
}

impl<N, W> Hash for Arc<N, W> {
    fn hash<H: Hasher>(&self, state: &mut H) { self.id.hash(state); }
}
impl<N, W> PartialEq for Arc<N, W> {
    fn eq(&self, other: &Self) -> bool { self.id == other.id }
}
impl<N, W> Eq for Arc<N, W> {}

/// Abstract Base Trait for Graphs.
pub trait Graph<N, W> {
    fn add_node_by_value(&mut self, value: N) -> Rc<Node<N>>;
    fn add_node(&mut self, node: Rc<Node<N>>);

    fn add_arc(&mut self, source: Rc<Node<N>>, target: Rc<Node<N>>, weight: W) -> Rc<Arc<N, W>> where W: Clone + 'static;
    fn add_unweighted_arc(&mut self, source: Rc<Node<N>>, target: Rc<Node<N>>) -> Rc<Arc<N, W>>;
    fn add_arc_func(&mut self, source: Rc<Node<N>>, target: Rc<Node<N>>, weight_func: Box<dyn Fn() -> W>) -> Rc<Arc<N, W>>;

    fn print_graph(&self) where N: Display, W: Display;
    fn to_dot_file(&self, filename: &str) -> std::io::Result<()> where N: Display, W: Display;

    fn get_nodes(&self) -> &HashSet<Rc<Node<N>>>;
    fn get_arcs(&self) -> &HashSet<Rc<Arc<N, W>>>;

    fn get_outgoing_arcs(&self, node: &Rc<Node<N>>) -> Vec<Rc<Arc<N, W>>>;
    fn get_incoming_arcs(&self, node: &Rc<Node<N>>) -> Vec<Rc<Arc<N, W>>>;

    fn get_out_degree(&self, node: &Rc<Node<N>>) -> usize { self.get_outgoing_arcs(node).len() }
    fn get_in_degree(&self, node: &Rc<Node<N>>) -> usize { self.get_incoming_arcs(node).len() }

    fn contains(&self, node: &Rc<Node<N>>) -> bool {
        self.get_nodes().contains(node)
    }

    /// Persists the graph's current state into a text file.
    fn save_to_file(&self, filename: &str) -> std::io::Result<()> where N: Display, W: Display {
        let mut file = File::create(filename)?;
        writeln!(file, "{}", self.get_nodes().len())?;

        let mut node_to_index = HashMap::new();
        let ordered_nodes: Vec<_> = self.get_nodes().iter().cloned().collect();

        for (index, node) in ordered_nodes.iter().enumerate() {
            node_to_index.insert(node.clone(), index);
            writeln!(file, "{}", node.value)?;
        }

        writeln!(file, "{}", self.get_arcs().len())?;

        for arc in self.get_arcs() {
            let src_idx = node_to_index[&arc.source];
            let tgt_idx = node_to_index[&arc.target];
            if let Some(weight) = arc.get_weight() {
                writeln!(file, "{} {} {}", src_idx, tgt_idx, weight)?;
            } else {
                writeln!(file, "{} {}", src_idx, tgt_idx)?;
            }
        }
        Ok(())
    }
}

/// Implementation of a Directed Graph.
pub struct DirectedGraph<N, W> {
    pub nodes: HashSet<Rc<Node<N>>>,
    pub arcs: HashSet<Rc<Arc<N, W>>>,
    out_adj_list: HashMap<Rc<Node<N>>, Vec<Rc<Arc<N, W>>>>,
    in_adj_list: HashMap<Rc<Node<N>>, Vec<Rc<Arc<N, W>>>>,
}

impl<N, W> DirectedGraph<N, W> {
    pub fn new() -> Self {
        Self { nodes: HashSet::new(), arcs: HashSet::new(), out_adj_list: HashMap::new(), in_adj_list: HashMap::new() }
    }

    pub fn load_from_file(filename: &str) -> std::io::Result<Self> where N: FromStr, W: FromStr + Clone + 'static {
        let mut file = File::open(filename)?;
        let mut content = String::new();
        file.read_to_string(&mut content)?;
        let tokens: Vec<String> = content.split_whitespace().map(|s| s.to_string()).collect();

        let mut g = Self::new();
        if tokens.is_empty() { return Ok(g); }

        let mut token_idx = 0;
        let num_nodes: usize = tokens[token_idx].parse().unwrap_or(0); token_idx += 1;

        let mut index_to_node = Vec::new();
        for _ in 0..num_nodes {
            if let Ok(val) = tokens[token_idx].parse::<N>() { index_to_node.push(g.add_node_by_value(val)); }
            token_idx += 1;
        }

        if token_idx >= tokens.len() { return Ok(g); }
        let num_arcs: usize = tokens[token_idx].parse().unwrap_or(0); token_idx += 1;
        let tokens_per_arc = if num_arcs > 0 { (tokens.len() - token_idx) / num_arcs } else { 0 };

        for _ in 0..num_arcs {
            let src_idx: usize = tokens[token_idx].parse().unwrap();
            let tgt_idx: usize = tokens[token_idx + 1].parse().unwrap();
            if tokens_per_arc >= 3 {
                if let Ok(weight) = tokens[token_idx + 2].parse::<W>() { g.add_arc(index_to_node[src_idx].clone(), index_to_node[tgt_idx].clone(), weight); }
            } else {
                g.add_unweighted_arc(index_to_node[src_idx].clone(), index_to_node[tgt_idx].clone());
            }
            token_idx += tokens_per_arc;
        }
        Ok(g)
    }
}

impl<N, W> Default for DirectedGraph<N, W> {
    fn default() -> Self {
        Self::new()
    }
}

impl<N, W> Graph<N, W> for DirectedGraph<N, W> {
    fn add_node_by_value(&mut self, value: N) -> Rc<Node<N>> {
        let node = Rc::new(Node::new(value));
        self.nodes.insert(node.clone());
        self.out_adj_list.insert(node.clone(), Vec::new());
        self.in_adj_list.insert(node.clone(), Vec::new());
        node
    }

    fn add_node(&mut self, node: Rc<Node<N>>) {
        self.nodes.insert(node.clone());
        self.out_adj_list.entry(node.clone()).or_insert_with(Vec::new);
        self.in_adj_list.entry(node.clone()).or_insert_with(Vec::new);
    }

    fn add_arc(&mut self, source: Rc<Node<N>>, target: Rc<Node<N>>, weight: W) -> Rc<Arc<N, W>> where W: Clone + 'static {
        let arc = Rc::new(Arc::new_fixed(source.clone(), target.clone(), weight));
        self.arcs.insert(arc.clone());
        self.out_adj_list.get_mut(&source).unwrap().push(arc.clone());
        self.in_adj_list.get_mut(&target).unwrap().push(arc.clone());
        arc
    }

    fn add_unweighted_arc(&mut self, source: Rc<Node<N>>, target: Rc<Node<N>>) -> Rc<Arc<N, W>> {
        let arc = Rc::new(Arc::new_unweighted(source.clone(), target.clone()));
        self.arcs.insert(arc.clone());
        self.out_adj_list.get_mut(&source).unwrap().push(arc.clone());
        self.in_adj_list.get_mut(&target).unwrap().push(arc.clone());
        arc
    }

    fn add_arc_func(&mut self, source: Rc<Node<N>>, target: Rc<Node<N>>, weight_func: Box<dyn Fn() -> W>) -> Rc<Arc<N, W>> {
        let arc = Rc::new(Arc::new(source.clone(), target.clone(), Some(weight_func)));
        self.arcs.insert(arc.clone());
        self.out_adj_list.get_mut(&source).unwrap().push(arc.clone());
        self.in_adj_list.get_mut(&target).unwrap().push(arc.clone());
        arc
    }

    fn print_graph(&self) where N: Display, W: Display { for arc in &self.arcs { arc.print(true); } }
    fn to_dot_file(&self, filename: &str) -> std::io::Result<()> where N: Display, W: Display {
        let mut file = File::create(filename)?;
        writeln!(file, "digraph G {{")?;
        for node in &self.nodes { writeln!(file, "  \"{}\";", node.value)?; }
        for arc in &self.arcs {
            if let Some(w) = arc.get_weight() { writeln!(file, "  \"{}\" -> \"{}\" [label=\"{}\"];", arc.source.value, arc.target.value, w)?; }
            else { writeln!(file, "  \"{}\" -> \"{}\";", arc.source.value, arc.target.value)?; }
        }
        writeln!(file, "}}")?;
        Ok(())
    }

    fn get_nodes(&self) -> &HashSet<Rc<Node<N>>> { &self.nodes }
    fn get_arcs(&self) -> &HashSet<Rc<Arc<N, W>>> { &self.arcs }
    fn get_outgoing_arcs(&self, node: &Rc<Node<N>>) -> Vec<Rc<Arc<N, W>>> { self.out_adj_list.get(node).cloned().unwrap_or_default() }
    fn get_incoming_arcs(&self, node: &Rc<Node<N>>) -> Vec<Rc<Arc<N, W>>> { self.in_adj_list.get(node).cloned().unwrap_or_default() }
}

/// Implementation of an Undirected Graph.
pub struct UndirectedGraph<N, W> {
    pub nodes: HashSet<Rc<Node<N>>>,
    pub arcs: HashSet<Rc<Arc<N, W>>>,
    out_adj_list: HashMap<Rc<Node<N>>, Vec<Rc<Arc<N, W>>>>,
    in_adj_list: HashMap<Rc<Node<N>>, Vec<Rc<Arc<N, W>>>>,
}

impl<N, W> UndirectedGraph<N, W> {
    pub fn new() -> Self {
        Self { nodes: HashSet::new(), arcs: HashSet::new(), out_adj_list: HashMap::new(), in_adj_list: HashMap::new() }
    }

    pub fn load_from_file(filename: &str) -> std::io::Result<Self> where N: FromStr, W: FromStr + Clone + 'static {
        let mut file = File::open(filename)?;
        let mut content = String::new();
        file.read_to_string(&mut content)?;
        let tokens: Vec<String> = content.split_whitespace().map(|s| s.to_string()).collect();

        let mut g = Self::new();
        if tokens.is_empty() { return Ok(g); }

        let mut token_idx = 0;
        let num_nodes: usize = tokens[token_idx].parse().unwrap_or(0); token_idx += 1;

        let mut index_to_node = Vec::new();
        for _ in 0..num_nodes {
            if let Ok(val) = tokens[token_idx].parse::<N>() { index_to_node.push(g.add_node_by_value(val)); }
            token_idx += 1;
        }

        if token_idx >= tokens.len() { return Ok(g); }
        let num_arcs: usize = tokens[token_idx].parse().unwrap_or(0); token_idx += 1;
        let tokens_per_arc = if num_arcs > 0 { (tokens.len() - token_idx) / num_arcs } else { 0 };

        for _ in 0..num_arcs {
            let src_idx: usize = tokens[token_idx].parse().unwrap();
            let tgt_idx: usize = tokens[token_idx + 1].parse().unwrap();
            if tokens_per_arc >= 3 {
                if let Ok(weight) = tokens[token_idx + 2].parse::<W>() { g.add_arc(index_to_node[src_idx].clone(), index_to_node[tgt_idx].clone(), weight); }
            } else {
                g.add_unweighted_arc(index_to_node[src_idx].clone(), index_to_node[tgt_idx].clone());
            }
            token_idx += tokens_per_arc;
        }
        Ok(g)
    }
}

impl<N, W> Default for UndirectedGraph<N, W> {
    fn default() -> Self {
        Self::new()
    }
}

impl<N, W> Graph<N, W> for UndirectedGraph<N, W> {
    fn add_node_by_value(&mut self, value: N) -> Rc<Node<N>> {
        let node = Rc::new(Node::new(value));
        self.nodes.insert(node.clone());
        self.out_adj_list.insert(node.clone(), Vec::new());
        self.in_adj_list.insert(node.clone(), Vec::new());
        node
    }

    fn add_node(&mut self, node: Rc<Node<N>>) {
        self.nodes.insert(node.clone());
        self.out_adj_list.entry(node.clone()).or_insert_with(Vec::new);
        self.in_adj_list.entry(node.clone()).or_insert_with(Vec::new);
    }

    fn add_arc(&mut self, source: Rc<Node<N>>, target: Rc<Node<N>>, weight: W) -> Rc<Arc<N, W>> where W: Clone + 'static {
        let arc = Rc::new(Arc::new_fixed(source.clone(), target.clone(), weight));
        self.arcs.insert(arc.clone());
        self.out_adj_list.get_mut(&source).unwrap().push(arc.clone());
        self.in_adj_list.get_mut(&source).unwrap().push(arc.clone());
        self.out_adj_list.get_mut(&target).unwrap().push(arc.clone());
        self.in_adj_list.get_mut(&target).unwrap().push(arc.clone());
        arc
    }

    fn add_unweighted_arc(&mut self, source: Rc<Node<N>>, target: Rc<Node<N>>) -> Rc<Arc<N, W>> {
        let arc = Rc::new(Arc::new_unweighted(source.clone(), target.clone()));
        self.arcs.insert(arc.clone());
        self.out_adj_list.get_mut(&source).unwrap().push(arc.clone());
        self.in_adj_list.get_mut(&source).unwrap().push(arc.clone());
        self.out_adj_list.get_mut(&target).unwrap().push(arc.clone());
        self.in_adj_list.get_mut(&target).unwrap().push(arc.clone());
        arc
    }

    fn add_arc_func(&mut self, source: Rc<Node<N>>, target: Rc<Node<N>>, weight_func: Box<dyn Fn() -> W>) -> Rc<Arc<N, W>> {
        let arc = Rc::new(Arc::new(source.clone(), target.clone(), Some(weight_func)));
        self.arcs.insert(arc.clone());
        self.out_adj_list.get_mut(&source).unwrap().push(arc.clone());
        self.in_adj_list.get_mut(&source).unwrap().push(arc.clone());
        self.out_adj_list.get_mut(&target).unwrap().push(arc.clone());
        self.in_adj_list.get_mut(&target).unwrap().push(arc.clone());
        arc
    }

    fn print_graph(&self) where N: Display, W: Display { for arc in &self.arcs { arc.print(false); } }
    fn to_dot_file(&self, filename: &str) -> std::io::Result<()> where N: Display, W: Display {
        let mut file = File::create(filename)?;
        writeln!(file, "graph G {{")?;
        for node in &self.nodes { writeln!(file, "  \"{}\";", node.value)?; }
        for arc in &self.arcs {
            if let Some(w) = arc.get_weight() { writeln!(file, "  \"{}\" -- \"{}\" [label=\"{}\"];", arc.source.value, arc.target.value, w)?; }
            else { writeln!(file, "  \"{}\" -- \"{}\";", arc.source.value, arc.target.value)?; }
        }
        writeln!(file, "}}")?;
        Ok(())
    }

    fn get_nodes(&self) -> &HashSet<Rc<Node<N>>> { &self.nodes }
    fn get_arcs(&self) -> &HashSet<Rc<Arc<N, W>>> { &self.arcs }
    fn get_outgoing_arcs(&self, node: &Rc<Node<N>>) -> Vec<Rc<Arc<N, W>>> { self.out_adj_list.get(node).cloned().unwrap_or_default() }
    fn get_incoming_arcs(&self, node: &Rc<Node<N>>) -> Vec<Rc<Arc<N, W>>> { self.in_adj_list.get(node).cloned().unwrap_or_default() }
}